gobblin-config-management/src/main/java/gobblin/config/client/ConfigClient.java

```diff
@@ -0,0 +1,273 @@
+    if (PathUtils.checkDescendant(floorKey, uri)) {
+      return this.configStoreMap.get(floorKey);
+    }
+
+    ConfigStoreFactory<ConfigStore> csFactory = this.getConfigStoreFactory(uri);
+    ConfigStore cs = csFactory.createConfigStore(uri);
+
+    if (!(cs instanceof ConfigStoreWithStableVersion)) {
+      if (this.policy == VERSION_STABILITY_POLICY.RAISE_ERROR) {
+        throw new Exception(String.format("Try to connect to unstable config store ", cs.getStoreURI()));
+      }
+    }
+
+    URI key = cs.getStoreURI();
+    ConfigStoreAccessor value = new ConfigStoreAccessor(cs, cs.getCurrentVersion());
+
+    // create default resolver
+    if (!(cs instanceof ConfigStoreWithResolution)) {
+      SimpleConfigStoreResolver resolver = new SimpleConfigStoreResolver(cs);
+      value.resolver = resolver;
+    }
+
+    // create default importedBy mapping object
+    if (!((cs instanceof ConfigStoreWithImportedBy) && (cs instanceof ConfigStoreWithImportedByRecursively))) {
+      SimpleImportMappings im = null;
+      // need to create the SimpleImportMappings using the resolver NOT the raw configStore
+      // otherwise, the getImportsRecursively and getImportedByRecursively will not work
+      // as raw configstore is NOT ConfigStoreWithResolution
+      if (!(cs instanceof ConfigStoreWithResolution)) {
+        im = new SimpleImportMappings(value.resolver, value.version);
+      } else {
+        im = new SimpleImportMappings(cs, value.version);
+      }
+      value.simpleImportMappings = im;
+    }
+
+    this.configStoreMap.put(key, value);
+    LOG.info(String.format("Created new config store with uri: %s, version: %s", cs.getStoreURI(),
+        cs.getCurrentVersion()));
+    return value;
+  }
+
+  // use serviceLoader to load configStoreFactories
+  @SuppressWarnings("unchecked")
+  private ConfigStoreFactory<ConfigStore> getConfigStoreFactory(URI uri) throws Exception {
+    ConfigStoreFactory csf = configStoreFactoryMap.get(uri.getScheme());
+    if (csf == null) {
+      throw new Exception("can not find corresponding config store factory for scheme " + uri.getScheme());
+    }
+
+    return (ConfigStoreFactory<ConfigStore>) csf;
+  }
+
+  /**
+   * 
+   * @param uri - must start with scheme name
+   * @return - the directly and indirectly specified configuration in {@com.typesafe.config.Config} format for input uri 
+   * 
+   * <p>
+   * detail logics:
+   * 1. Used the ConfigStoreAccessor in the cached map if previous ConfigStore been queried
+   * 2. Based the scheme name, using {@java.util.ServiceLoader} to find the first {@gobblin.config.configstore.ConfigStoreFactory}
+   * 3. Use ConfigStoreFactory to create ConfigStore 
+   * 4. Build ConfigStoreAccessor by checking the current version of the ConfigStore. Added the entry to theMap
+   * 5. If the ConfigStore is NOT ConfigStoreWithResolution, need to do resolution in this client
+   */
+  public Config getConfig(URI uri) throws Exception {
```

Comment:
Let's also avoid generic names as "uri" as they convey very little information. Let's, for example, agree to call the config URIs, configKeyUri if it is a URI or configKey if it is a String.